### PR TITLE
Add BBR Example Test Cases Appendix

### LINKED ISSUE #71: Consider adding test cases to an appendix
Today in CCWG there was interest in example test cases of some form for BBR, so implementers could ensure they implemented it correctly.

QLog was mentioned as a common logging framework for QUIC that could be used for congestion control debugging and might allow implementations to compare traces.

### PR DESCRIPTION
**Pull request by ianswett**
This PR fixes https://github.com/ietf-wg-ccwg/draft-ietf-ccwg-bbr/issues/71 by adding an Appendix section titled "BBR Example Test Cases". It details the 16 test cases originally discussed, along with 4 new edge cases brainstormed during planning. The cases use step-by-step logic detailing state transitions and variables without relying on QLog text.

---
*PR created automatically by Jules for task [4897439197073774623](https://jules.google.com/task/4897439197073774623) started by @ianswett*

### CODE REVIEW

File: draft-ietf-ccwg-bbr.md [hunk 1]
```diff
@@ -3621,3 +3621,231 @@ for feedback, suggestions, and edits on earlier versions of this document.
...
+To ensure internal state variables, including the Bandwidth Estimator,
+are initialized correctly, when an test case says it starts in a phase
+after STARTUP, it is best to start a flow and drive it so it exits STARTUP
+and naturally gets to the specified phase.
+
+## Exiting STARTUP on Bandwidth Plateau
+
+1. Start a flow in the STARTUP phase.
+2. Ensure the sender has enough data to send and is not application-limited.
+3. Observe that BBR measures delivery rate samples that plateau.
```
Comment (ianswett) [inline draft-ietf-ccwg-bbr.md]: And ensure the measured bandwidth is close (ie: within 2%) to the simulated bandwidth.

File: draft-ietf-ccwg-bbr.md [hunk 2]
```diff
@@ -3621,3 +3621,231 @@ for feedback, suggestions, and edits on earlier versions of this document.
...
+2. Suddenly increase the bottleneck bandwidth by 10x (e.g., to 100 Mbps).
+3. Verify that during PROBE_UP, the additive increase to `BBR.inflight_longterm`
+   exponentially doubles each round trip.
+4. Ensure the connection discovers the full 100 Mbps bandwidth within a
+   reasonable number of round trips (O(log(BDP))).
+
+## Decrease Bandwidth 10x and ensure Max Bandwidth adapts down
+
+1. Wait until the connection transitions to PROBE_BW.
+2. Suddenly decrease the bottleneck bandwidth by 10x (e.g., to 10 Mbps).
```
Comment (ianswett) [inline draft-ietf-ccwg-bbr.md]:
```suggestion
2. Suddenly decrease the bottleneck bandwidth by 10x (e.g., to 1 Mbps).
```

File: draft-ietf-ccwg-bbr.md [hunk 3]
```diff
@@ -3621,3 +3621,231 @@ for feedback, suggestions, and edits on earlier versions of this document.
...
+3. Verify that `BBR.bw_shortterm` rapidly adapts downwards due to the resulting
+   packet loss.
+4. Verify that after the `BBR.max_bw_filter` window expires (2 PROBE_BW cycles),
+   `BBR.max_bw` drops to the new 10 Mbps limit, matching the path's delivery
+   rate.
+
+## Token Bucket Policer
+
+1. Configure the path with a token bucket policer (e.g., rate limit with a
+   specific initial burst size and no additional buffering beyond the bucket).
```
Comment (nealcardwell) [inline draft-ietf-ccwg-bbr.md:3819]: Let's cite the 2016 sigcomm paper on policers:

An Internet-Wide Analysis of Traffic Policing
https://dl.acm.org/doi/10.1145/2934872.2934873

File: draft-ietf-ccwg-bbr.md [hunk 4]
```diff
@@ -3621,3 +3621,227 @@ for feedback, suggestions, and edits on earlier versions of this document.
...
 --- back
+
+# BBR Example Test Cases {#bbr-example-test-cases}
+
+This section describes some test cases to ensure BBR implementations perform as
+expected. When they are important to the test case, variables from the
+pseudocode are referenced. Most test cases focus on state transitions, since
+those are critical to BBR's overall behavior.
+
+This assumes that the packet pacer and the Bandwidth Estimator are implemented
```
Comment (nealcardwell) [inline draft-ietf-ccwg-bbr.md]: Maybe we can remove  this sentence: "This assumes that the packet pacer and the Bandwidth Estimator are implemented correctly." 

And instead we can have the first test case check that the pacer and bw estimator are working correctly by verifying the values of RS.delivery_rate and BBR.max_bw match the expected values.

File: draft-ietf-ccwg-bbr.md [hunk 5]
```diff
@@ -3621,3 +3621,227 @@ for feedback, suggestions, and edits on earlier versions of this document.
...
+correctly.
+
+## Exiting STARTUP on Bandwidth Plateau
+
+1. Start a flow in the STARTUP phase.
+2. Ensure the sender has enough data to send and is not application-limited.
+3. After several round trips, artificially cap the bottleneck bandwidth to a
```
Comment (nealcardwell) [inline draft-ietf-ccwg-bbr.md]: Rather than phrasing this as "After several round trips, artificially cap the bottleneck bandwidth", I would suggest that it would be simpler to think about and implement if most cases test a fixed bandwidth for the entire test. So perhaps the text can clarify up front, before any of the test cases, something like:

"Unless otherwise mentioned, each test case is configured to create a single BBR connection traveling over a controlled path with a bottleneck bandwidth and two-way propagation delay that are fixed for the entire duration of the test. The tests can use network emulation, network simulation, or a controlled network path such as a laboratory environment."

File: draft-ietf-ccwg-bbr.md [hunk 6]
```diff
@@ -3621,3 +3621,227 @@ for feedback, suggestions, and edits on earlier versions of this document.
...
+   constant rate (e.g., 10 Mbps).
+4. Observe that BBR measures delivery rate samples that plateau, and ensure the
```
Comment (nealcardwell) [inline draft-ietf-ccwg-bbr.md]: I would suggest explicitly calling for verifying the RS.delivery_rate and BBR.max_bw; something like:

4. Verify that BBR measures delivery rate samples (RS.delivery_rate) that plateau, and ensure the estimated bandwidth (BBR.max_bw) is close (i.e., within 2%) to the simulated bandwidth.

File: draft-ietf-ccwg-bbr.md [hunk 7]
```diff
@@ -3621,3 +3621,227 @@ for feedback, suggestions, and edits on earlier versions of this document.
...
+
+# BBR Example Test Cases {#bbr-example-test-cases}
+
+This section describes some test cases to ensure BBR implementations perform as
+expected. When they are important to the test case, variables from the
+pseudocode are referenced. Most test cases focus on state transitions, since
+those are critical to BBR's overall behavior.
+
+This assumes that the packet pacer and the Bandwidth Estimator are implemented
+correctly.
```
Comment (ianswett) [inline draft-ietf-ccwg-bbr.md]:
```suggestion
The examples have checks to ensure the Bandwidth Estimator and pacer
are meeting BBR's needs, but there are no standalone tests for either.
```

File: draft-ietf-ccwg-bbr.md (hunk 4 above)
Comment (ianswett) [reply draft-ietf-ccwg-bbr.md]: I changed it to:
"The examples have checks to ensure the Bandwidth Estimator and pacer
are meeting BBR's needs, but there are no standalone tests for either."

Let me know if that's better?

File: draft-ietf-ccwg-bbr.md [hunk 8]
```diff
@@ -3621,3 +3621,231 @@ for feedback, suggestions, and edits on earlier versions of this document.
...
```
Comment (ianswett) [inline draft-ietf-ccwg-bbr.md]:
```suggestion
The examples have checks to ensure the Bandwidth Estimator and pacer
are meeting BBR's needs, but there are no standalone tests for either.
```

File: draft-ietf-ccwg-bbr.md [hunk 9]
```diff
@@ -3621,3 +3621,231 @@ for feedback, suggestions, and edits on earlier versions of this document.
...
+To ensure internal state variables, including the Bandwidth Estimator,
+are initialized correctly, when an test case says it starts in a phase
+after STARTUP, it is best to start a flow and drive it so it exits STARTUP
+and naturally gets to the specified phase.
+
+## Exiting STARTUP on Bandwidth Plateau
+
+1. Start a flow in the STARTUP phase.
+2. Ensure the sender has enough data to send and is not application-limited.
+3. Observe that BBR measures delivery rate samples that plateau.
```
Comment (ianswett) [inline draft-ietf-ccwg-bbr.md]:
```suggestion
3. Observe that BBR measures delivery rate samples that plateau and are
BBR.bw_latest is within 2% of the simulated bandwidth.
```

Comment (ianswett) [inline draft-ietf-ccwg-bbr.md]:
```suggestion
3. Observe that BBR measures delivery rate samples that plateau and are
BBR.bw_latest is within 2% of the simulated bandwidth.
```

File: draft-ietf-ccwg-bbr.md [hunk 10]
```diff
@@ -3621,3 +3625,233 @@ for feedback, suggestions, and edits on earlier versions of this document.
...
+are initialized correctly, when an test case says it starts in a phase
+after STARTUP, it is best to start a flow and drive it so it exits STARTUP
+and naturally gets to the specified phase.
+
+## Exiting STARTUP on Bandwidth Plateau
+
+1. Start a flow in the STARTUP phase.
+2. Ensure the sender has enough data to send and is not application-limited.
+3. Observe that BBR measures delivery rate samples that plateau and are
+BBR.bw_latest is within 2% of the simulated bandwidth.
```
Comment (ianswett) [inline draft-ietf-ccwg-bbr.md]: Neal's suggestion

```suggestion
3. Verify that BBR measures delivery rate samples (RS.delivery_rate) that plateau,
and ensure the estimated bandwidth (BBR.max_bw) is close (i.e., within 2%) to
the simulated bandwidth.
```

File: draft-ietf-ccwg-bbr.md (hunk 6 above)
Comment (ianswett) [reply draft-ietf-ccwg-bbr.md]: Done

File: draft-ietf-ccwg-bbr.md [hunk 11]
```diff
@@ -3621,3 +3625,234 @@ for feedback, suggestions, and edits on earlier versions of this document.
...
+
+The examples have checks to ensure the Bandwidth Estimator and pacer
+are meeting BBR's needs, but there are no standalone tests for either.
+
+Examples are assumed to have a fixed bandwidth (ie: 10Mbps) and a large enough
+buffer that losses caused by buffer overflow will not occur, unless specified.
+To ensure internal state variables, including the Bandwidth Estimator,
+are initialized correctly, when an test case says it starts in a phase
+after STARTUP, it is best to start a flow and drive it so it exits STARTUP
+and naturally gets to the specified phase.
```
Comment (ianswett) [inline draft-ietf-ccwg-bbr.md]:
```suggestion
Unless otherwise mentioned, each test case is configured to create a single BBR connection traveling over a controlled path with a bottleneck bandwidth (ie: 10Mbps) and two-way propagation delay that are fixed for the entire duration of the test. The tests can use network emulation, network simulation, or a controlled network path such as a laboratory environment.

To ensure internal state variables, including the Bandwidth Estimator,
are initialized correctly, when an test case says it starts in a phase
after STARTUP, test cases expect the connection to be driven with enough
data it so it exits STARTUP and naturally gets to the specified phase.
```

File: draft-ietf-ccwg-bbr.md (hunk 5 above)
Comment (ianswett) [reply draft-ietf-ccwg-bbr.md]: Thanks, I worked that into the intro above all the examples

File: draft-ietf-ccwg-bbr.md [hunk 12]
```diff
@@ -3621,3 +3625,237 @@ for feedback, suggestions, and edits on earlier versions of this document.
...
+   Recovery or RTO.
+2. Verify `BBRSaveStateUponLoss()` saves the current `BBR.state`,
+   `BBR.bw_shortterm`, `BBR.inflight_shortterm`, and `BBR.inflight_longterm`.
+3. Once the transport protocol detects the loss was spurious (e.g., receiving an
+   ACK for the retransmitted packet that shows it was already delivered), verify
+   `BBRHandleSpuriousLossDetection()` restores the saved parameters.
+4. Ensure the connection seamlessly returns to its previous state (e.g.,
+   PROBE_UP).
+
+## Entering and Exiting PROBE_RTT during STARTUP
```
Comment (ianswett) [inline draft-ietf-ccwg-bbr.md]:
```suggestion
## Entering and Exiting PROBE_RTT during STARTUP
```

File: draft-ietf-ccwg-bbr.md [hunk 13]
```diff
@@ -3621,3 +3625,237 @@ for feedback, suggestions, and edits on earlier versions of this document.
...
+
+1. Configure a bottleneck with a very low BDP (e.g., < 1 packet).
+2. Wait until the connection transitions to PROBE_BW.
+3. Verify that `BBR.MinPipeCwnd` (4 packets) acts as a floor for `C.cwnd`.
+4. Ensure the pacing rate correctly matches the low bandwidth, while the
+   4-packet cwnd allows for delayed ACKs without stalling the pipeline.
+
+## Increase Bandwidth 10x and ensure full bandwidth is reached
+
+1. Wait until the connection transitions to PROBE_BW.
```
Comment (antoniovicente) [inline draft-ietf-ccwg-bbr.md:3799]: Would it make sense to constrain this further by targeting an increase during either probebw_down or probebw_cruise?

File: draft-ietf-ccwg-bbr.md [hunk 14]
```diff
@@ -3621,3 +3625,237 @@ for feedback, suggestions, and edits on earlier versions of this document.
...
+2. Suddenly increase the bottleneck bandwidth by 10x (e.g., to 100 Mbps).
+3. Verify that during PROBE_UP, the additive increase to `BBR.inflight_longterm`
+   exponentially doubles each round trip.
+4. Ensure the connection discovers the full 100 Mbps bandwidth within a
+   reasonable number of round trips (O(log(BDP))).
```
Comment (antoniovicente) [inline draft-ietf-ccwg-bbr.md:3804]: Is a tighter bound possible like `N * log(increase) / log(pacing_gain) + C`

For this specific case of 10x increase it is expected that the new bandwidth be discovered within ~15 RTTs (for increase=10, pacing_gain=1.25, N=1.0 and C=4)?

File: draft-ietf-ccwg-bbr.md [hunk 15]
```diff
@@ -3621,3 +3625,237 @@ for feedback, suggestions, and edits on earlier versions of this document.
...
+1. Start a flow in the STARTUP phase.
+2. Ensure the sender has enough data to send and is not application-limited.
+3. Verify that BBR measures delivery rate samples (RS.delivery_rate) that plateau,
+   and ensure the estimated bandwidth (BBR.max_bw) is close (i.e., within 2%) to
+   the simulated bandwidth.
+4. Verify that after 3 consecutive rounds where the delivery rate grows by less
+   than 25%, `BBR.full_bw_now` and `BBR.full_bw_reached` are set to `true`.
+5. Verify the connection transitions from STARTUP to DRAIN.
+
+## Exiting STARTUP on Loss when app-limited
```
Comment (antoniovicente) [inline draft-ietf-ccwg-bbr.md:3664]: Another interesting case: `No exit from STARTUP when app-limited and there is no loss.`

**Review by antoniovicente — COMMENTED** [review]
Thanks for sending me a link to the list of test cases.  I did an initial read and need to go back to look at them in more detail but it seems like an excellent improvement.  It's good to have concrete cases to test again which cover the main design concepts behind BBR and operational aspects of the different phases.

File: draft-ietf-ccwg-bbr.md [hunk 16]
```diff
@@ -3621,3 +3625,237 @@ for feedback, suggestions, and edits on earlier versions of this document.
...
+as expected. When they are important to the test case, variables from the
+pseudocode are referenced. Most test cases focus on state transitions, since
+those are critical to BBR's overall behavior.
+
+The examples have checks to ensure the Bandwidth Estimator and pacer
+are meeting BBR's needs, but there are no standalone tests for either.
+
+Unless otherwise mentioned, each test case is configured to create a single
+BBR connection traveling over a controlled path with a bottleneck bandwidth
+(ie: 10Mbps) and two-way propagation delay that are fixed for the entire
```
Comment (antoniovicente) [inline draft-ietf-ccwg-bbr.md]: I really appreciate the tests taking RTT and bottleneck bandwidth as parameters.  Running the same test under small, medium and larger RTTs are likely to expose different problems.

Consider adding some wording encouraging implementers to explore behavior over a range of network params, including RTT=0 and 100msec+

File: draft-ietf-ccwg-bbr.md (hunk 15 above)
Comment (ianswett) [reply draft-ietf-ccwg-bbr.md:3664]: I added that with No exit from PROBE_UP when app-limited and there is no loss.

File: draft-ietf-ccwg-bbr.md (hunk 13 above)
Comment (ianswett) [reply draft-ietf-ccwg-bbr.md:3799]: Are you thinking to constrain the test more?

File: draft-ietf-ccwg-bbr.md (hunk 14 above)
Comment (ianswett) [reply draft-ietf-ccwg-bbr.md:3804]: Good point, we could put a tighter bound on it.  If inflight_longterm isn't set, it'd be about 11 Round trips ideally I believe? (1.25^11 > 10). But if inflight_longterm is set, it would depend upon the BDP.

File: draft-ietf-ccwg-bbr.md (hunk 16 above)
Comment (ianswett) [reply draft-ietf-ccwg-bbr.md]: Good point, I added a sentence on that.

**Review by nealcardwell — APPROVED** [review]
LGTM. Thanks!
